lib/exabgp/bgp/message/update/__init__.py

```diff
@@ -104,147 +104,72 @@ def split (data):
-						yield self._message(Update.prefix(packed_del) + Update.prefix(packed_mp_del + packed_mp_add + attr))
-						packed_del = ''
-						packed_mp_del = ''
-						packed_mp_add = packed
-					else:
-						packed_mp_add += packed
-			except StopIteration:
-				pass
-
-		# ADD Ipv4
-
-		packed_add = ''
-
-		if add_nlri:
-			msg_size = negotiated.msg_size - 19 - 2 - 2 - len(attr)  # 2 bytes for each of the two prefix() header
-		seen_size = len(packed_del + packed_mp_del + packed_mp_add)
-		if seen_size > msg_size:
-			yield self._message(Update.prefix(packed_del) + Update.prefix(packed_mp_del))
-			packed_del = ''
-			packed_mp_del = ''
-
-		# addpath = negotiated.addpath.send(AFI.ipv4,SAFI.unicast)
-
-		while add_nlri:
-			nlri = add_nlri.pop()
+		# Withdraws/NLRIS (IPv4 unicast and multicast)
+		msg_size = negotiated.msg_size - 19 - 2 - 2 - len(attr) # 2 bytes for each of the two prefix() header
+		withdraws = ''
+		announced = ''
+		for nlri in nlris:
 			packed = nlri.pack(negotiated)
-			seen_size = len(packed_del + packed_mp_del + packed_mp_add + packed_add + packed)
-			if seen_size > msg_size:
-				if not packed_add and not packed_mp_add and not packed_mp_del and not packed_del:
+			if len(announced + withdraws + packed) > msg_size:
+				if not withdraws and not announced:
 					raise Notify(6,0,'attributes size is so large we can not even pack one NLRI')
-				if packed_mp_add:
-					yield self._message(Update.prefix(packed_del) + Update.prefix(packed_mp_del + packed_mp_add + attr) + packed_add)
-					msg_size = negotiated.msg_size - 19 - 2 - 2  # 2 bytes for each of the two prefix() header
-				else:
-					yield self._message(Update.prefix(packed_del) + Update.prefix(packed_mp_del + attr) + packed_add)
-				packed_del = ''
-				packed_mp_del = ''
-				packed_mp_add = ''
-				packed_add = packed
+				yield self._message(Update.prefix(withdraws) + Update.prefix(attr) + announced)
+				if nlri.action == OUT.ANNOUNCE:
+					announced = packed
+					withdraws = ''
+				elif include_withdraw:
+					withdraws = packed
+					announced = ''
 			else:
-				packed_add += packed
-
-		yield self._message(Update.prefix(packed_del) + Update.prefix(packed_mp_del + packed_mp_add + attr) + packed_add)
+				if nlri.action == OUT.ANNOUNCE:
+					announced += packed
+				elif include_withdraw:
+					withdraws += packed
+
+		if mp_nlris:
+			for family in mp_nlris.keys():
+				afi, safi = family
+				mp_reach = ''
+				mp_unreach = ''
+				mp_announce = MPRNLRI(afi, safi, mp_nlris[family].get(OUT.ANNOUNCE, []))
+				mp_withdraw = MPURNLRI(afi, safi, mp_nlris[family].get(OUT.WITHDRAW, []))
+
+				for mprnlri in mp_announce.packed_attributes(negotiated, msg_size - len(withdraws + announced)):
+					if mp_reach:
+						yield self._message(Update.prefix(withdraws) + Update.prefix(attr + mp_reach) + announced)
+						announced = ''
+						withdraws = ''
+					mp_reach = mprnlri
+
+				if include_withdraw:
```

Comment:
If the path changes, you just need to send an UPDATE with the new path attributes. If the path really is gone, then yes, withdraw. On the receiving side the new UPDATE just overwrites the previous known path.